lib/hobbes/lang/preds/dependent.C

```diff
@@ -0,0 +1,588 @@
+    case 10:
+      return BUILD(_FixedArray);
+    case 11:
+      return BUILD(_TString);
+    case 12:
+      return BUILD(_TLong);
+    case 13:
+      return BUILD(_TVar);
+  }
+  return MonoTypePtr();
+}
+
+
+MonoTypePtr type_; // Filthy hack since I can't seem to get type def from typedb
+
+struct liftType {
+  static MonoTypePtr type(typedb&) {
+    return type_;
+  }
+};
+
+template <>
+  struct lift<Type*, false> : public liftType { };
+
+bool evalType(cc* ctx, const TEnvPtr& tenv, Definitions* ds, MonoTypePtr& tptr, ExprPtr ex, MonoTypePtr &r) {
+  try{
+
+    ExprPtr e = validateType(tenv, ex, ds);
+    if (!e) return false;
+    if (*e->type()->monoType() == *primty("long")) {
+      r = MonoTypePtr(TLong::make(ctx->compileFn<long()>(e)()));
+      return true;
+    } else if (*e->type()->monoType() == *arrayty(primty("char"))) {
+      array<char> * hs = ctx->compileFn<array<char>*()>(e)();
+      r = MonoTypePtr(TString::make(std::string(hs->data, hs->data + hs->size)));
+      return true;
+    } else if (*e->type()->monoType() == *tptr) {     
+      Type* te = ctx->compileFn<Type*()>(e)();
+      r = typeExprToMonoTypePtr(te);
+      return true;
+    } else {
+      return false;
+    }
+  } catch(...) { return false; }
+}
+
+
+bool decodeTAConstraint(cc*ctx, const TEnvPtr tenv, Definitions* ds, const ConstraintPtr& c, MonoTypePtr& l, MonoTypePtr& r) {
+  MonoTypePtr tptr = ctx->replaceTypeAliases(primty("Type"));
+  type_ = tptr;
+  LexicalAnnotation la;
+  if (c->name() == TypeApply::constraintName() && c->arguments().size() > 1) {
+    l = c->arguments()[0];
+    if (TApp * tapp = is<TApp>(c->arguments()[1])) {
+      if(tapp->args().size() == 1 && *tapp->fn() == *primty("quote")) {
+        if (TExpr* te = is<TExpr>(tapp->args()[0])) {
+          ExprPtr f = te->expr();
+          Exprs args;
+
+          for(unsigned int i = 2; i < c->arguments().size(); ++i) {
+            ExprPtr arg;
+	    MonoTypePtr targ = c->arguments()[i];
+	    if (is<TVar>(targ)) {
+              return false;
+	    }
+            if (TApp* tapp = is<TApp>(targ)) {
+              if (*tapp->fn() == *primty("quote")) {
+                if (TExpr* e = is<TExpr>(tapp->args()[0])) {
+                  arg = e->expr();
+                }
+              }
+            }
+	    if (TString* tstr = is<TString>(targ)) {
+              arg = ExprPtr(mkarray(tstr->value(), la));
+            }
+            if (TLong* tlong = is<TLong>(targ)) {
+              arg = constant(tlong->value(), la);
+            }
+
+            if(!arg) {arg = switchOf(targ, MonoTypePtrToExpr(tenv, tptr));}
+            if (arg) {
+              args.push_back(arg);
+            } else {
+              return false;
+            }
+          }
+
+          if(evalType(ctx, tenv, ds, tptr, fncall(f, args, la), r)) {
+            return true;
+          }
+          // This doesn't seem to work how I'd hoped, closures don't get eval'ed
+          // properly.
```

Comment:
What happens that you aren't expecting?

Reply:
I can't get this sort of thing

    t1 :: (TypeApply a `\x.\y.x+y` 3) => () -> a
    t1 = \().unsafeCast(())

    t2 :: (TypeApply b `\x y.apply` a 2) => a -> ((Z b))
    t2 = unsafeCast

to work, but actually it seems to be dying on the def of t1

Reply:
That's an interesting example, because in `t1` your result for `a` should actually have a qualified polymorphic type like `(Add int a b) => exists E.(((E,a)->b) * E)`.  However, type constraints only allow monomorphic types (or type variables) in their arguments, so it's kind of a problem of the result being "too big to fit".

That will be a problem for `TypeApply` with any such type (not just closures).

The fact that LLVM gets involved and we do some compilation to determine a result for one of these constraints is another way to get to the same conclusion.  We can only produce machine code (so only involve LLVM) at monomorphic types.

Reply:
Sorry, should have been more exact with signatures, replacing `(+)` with `ladd`, doesn't change matters, and the ladd closure is certainly compilable.

Reply:
I see, I guess you have an arity mismatch in `t1` (since `TypeApply` expects 4 arguments).  But it does seem like monomorphic closure types can't get through either:

```
> :t newPrim() :: (TypeApply a `\x y.\z.iadd(x,iadd(y,z))` 21 1) => a
stdin:1,28-52: Expected existential type in closure application
```

Reply:
`TypeApply` takes a variable number of arguments, `t1` should be fine there, since there is only one argument to the first function.  One problem is that I've managed to delete some code, `t1` as stands can only be evaluated to a `long`, `[char]` or `Type`, the fallback code that shoves it into a quote has gone walkabout.  If only I had tests  ;)

Your attempt won't work, because you are using `iadd` and feeding in `long`s, (this is slightly annoying that type level longs are written without the `L`, and type level `0L` must be written as `false`, as `0` gets turned to `void`

Reply:
Facepalm, of course, my mistake re: type-level longs ... that should be changed.  I was too happy with the product/sum deconstruction bottoming out in unit (for products) and void (for sums) so I could write `a*b*1` or `a+b+0`.  In retrospect, it wasn't worth it.  I think we can change `0` to a type-level number instead of `void`.

Also the syntax difference for term-level long versus type-level long, that's annoying (and just tripped me up).  If we do go all the way down the rabbit hole with dependent types, then we won't be able to distinguish syntax between levels, so will have to resolve this.  :T